example/persist.py

```diff
@@ -0,0 +1,223 @@
+if would not be persisted correctly. This could be resolved by correctly
+merging the pv.current().raw and value.raw appropriately in the post().
+"""
+
+import json
+import sqlite3
+import time
+
+from p4p import Value
+from p4p.nt.scalar import NTScalar
+from p4p.server import Server, ServerOperation
+from p4p.server.raw import Handler
+from p4p.server.thread import SharedPV
+
+
+class PersistHandler(Handler):
+    """
+    A handler that will allow simple persistence of values and timestamps using
+    SQLite3 across retarts.
+
+    It requires the open handler in order to restore values from a previous run.
+    It requires a post handler in order to persist values set within the program.
+    """
+
+    def __init__(self, pv_name: str, conn: sqlite3.Connection, open_restore=True):
+        self._conn = conn
+        self._pv_name = pv_name
+        self._open_restore = open_restore
+
+    def open(self, value: Value):
+        """
+        Restore values from the previous run, if possible.
+        If there isn't a previous value then we persist this initial value.
+        """
+
+        # If there is a value already in the database we always use that
+        # instead of the supplied initial value, unless the
+        # _open_restore flag indicates otherwise.
+        if not self._open_restore:
+            return
+
+        # We could, in theory, re-apply authentication here if we queried for
+        # that information and then did something with it!
+        res = self._conn.execute("SELECT * FROM pvs WHERE id=?", [self._pv_name])
+        query_val = res.fetchone()
+
+        if query_val is not None:
+            json_val = json.loads(query_val[1])
+
+            # Report on restored values and setter (reports None if not known)
+            print(
+                f"Will restore to {self._pv_name} value: {json_val['value']}, set by {query_val[2]}"
+            )
+
+            # Override initial value
+            value["value"] = json_val["value"]
+
+            # Override current timestamp (if any) with time the new value was set
+            value["timeStamp.secondsPastEpoch"] = json_val["timeStamp"][
+                "secondsPastEpoch"
+            ]
+            value["timeStamp.nanoseconds"] = json_val["timeStamp"]["nanoseconds"]
+        else:
+            # We are using an initial value so persist it
+            self._upsert(value)
+
+    def post(self, pv: SharedPV, value: Value):
+        """Update timestamp of the PV and store its new value."""
+        self._update_timestamp(value)
+
+        self._upsert(value)
+
+    def put(self, pv: SharedPV, op: ServerOperation):
+        """
+        Update timestamp of the PV and store its new value. Unlike the post()
+        we include the account and peer update in the data stored.
+        """
+        value = op.value().raw
+
+        self._update_timestamp(value)
+        self._upsert(value, op.account(), op.peer())
```

Comment:
Is there a better way to do this? Originally I was using `pv.post(op.value())` which is much neater, but discards the information about `op.account()` and `op.peer()`. Using `op.value().raw` feels brittle, and will discard any data apart from the value.